### ISSUE TITLE
Default Firefox to MV3

### ISSUE THREAD

**Issue by aklinker1** [wxt-dev/wxt#230]
### Feature Request

Before this can happen, the following changes are required:

- [x] ~~_MV3 out of experimental status_~~ v109 added MV3 support
- [x] #229 
- [x] Dev mode support: https://bugzilla.mozilla.org/show_bug.cgi?id=1864284

#### Is your feature request related to a bug?

#226

**Comment by aklinker1** [comment]
>[!WARNING]
>If you're trying to run `wxt -b firefox --mv3`, use `wxt build -b firefox --mv3` instead to build your extension for production and load the extension manually to test Firefox MV3 extensions.
>
> For now, I would recommend not using Firefox for development, or using MV2 (the default) instead.

**Comment by dvlden** [comment]
I have released my extension on Firefox add-ons for Desktop and Android as MV3.
Instantly regretted it.

- Requires users to give extension access to all websites. It's not automatically allowed as in MV2.
- Firefox Android really [lacks MV3 compatibility](https://extensionworkshop.com/documentation/develop/developing-extensions-for-firefox-for-android/)

<img width="777" alt="Screenshot 2024-01-19 at 00 01 16" src="https://github.com/wxt-dev/wxt/assets/3460368/7c22b915-d137-4c67-ba29-dbf9167ea968">

> Releasing soon MV2 for Firefox instead as it resolves both of the issues from above.

Leaving this as a note. Firefox is not ready, maybe in a few years.

**Comment by aklinker1** [comment]
Yeah, I just released my rewrite as MV3 extension and the permissions are weird... Probably more secure, but less user friendly, for sure. I'm not supporting Android, so that's not a problem for me.

I'm gonna close this issue for now so I don't have to keep looking at it.

**Comment by yunsii** [comment]
But https://wxt.dev/guide/multiple-browsers.html and development console have no information about this 😂 I spend so many time to debug why firefox with mv3 not work...

**Comment by yunsii** [comment]
> WXT should exit with an error message when you try and run `wxt -b firefox --mv3`
> 
> <img alt="Screenshot 2024-02-20 at 8 27 37 AM" width="1150" src="https://private-user-images.githubusercontent.com/10101283/306297406-1ed58f22-7148-499d-a0db-e05eda64340a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0NzcwNjgsIm5iZiI6MTcwODQ3Njc2OCwicGF0aCI6Ii8xMDEwMTI4My8zMDYyOTc0MDYtMWVkNThmMjItNzE0OC00OTlkLWEwZGItZTA1ZWRhNjQzNDBhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDAwNTI0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUxZWI0NjNjZDhhZjk4YjM3OGRjNjRkZDQ5NWZlM2QwMDBjMTllNDZhNDcxNzMzZWEwNWQwNWVmZWUwMjY2NTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cfajfTpUSWr5GpeqCyQyz3X5vidgbhruNdMMpwmu4U8">

Really? Why I do not encounter any error message?
![image](https://github.com/wxt-dev/wxt/assets/18096089/8a3a9915-4b61-4e1a-9263-e1bbe25d2ab9)

**Comment by aklinker1** [comment]
Huh 😖

**Comment by aklinker1** [comment]
@yunsii Can you add this log to your project? Inside `node_modules/wxt/dist/cli.js`, do a `ctrl+F` for "Dev mode does not support Firefox MV3" and add this log before the if statement:

<img width="733" alt="Screenshot 2024-02-21 at 7 10 00 AM" src="https://github.com/wxt-dev/wxt/assets/10101283/f6ef0a82-5813-433a-9908-ee0f612288a6">

```ts
console.log(wxt.config.browser, wxt.config.manifestVersion)
```

Then re-run the command that's failing and let me know what is logged?

**Comment by yunsii** [comment]
@aklinker1 It seems a bug with WSL?
![image](https://github.com/wxt-dev/wxt/assets/18096089/28cf1ee4-ef25-4730-a3e3-eb4a76885d95)
https://github.com/wxt-dev/wxt/blob/86b7162d99acbf6a9c92f959cffc60255cc30fe6/src/core/runners/wsl.ts#L8-L22

And why exit with an error message at beginning of `dev` command?

**Comment by aklinker1** [comment]
> It seems a bug with WSL?

@yunsii This is intentional. `web-ext` doesn't work well in WSL: https://github.com/mozilla/web-ext/issues/2108#issue-779469169

**Comment by yunsii** [comment]
@aklinker1 So how about exit with an error message at start of `dev` command?

**Comment by yunsii** [comment]
> Firefox MV3 dev mode (`wxt -b firefox --mv3`) does not work, see [bugzilla.mozilla.org/show_bug.cgi?id=1864284](https://bugzilla.mozilla.org/show_bug.cgi?id=1864284) for more details. You have a two options:
> 
> 1. Use `--mv2`, the default, instead when developing against firefox `wxt -b firefox --mv2`
> 2. Build for production, `wxt build -b firefox --mv3`, and load the extension manually into Firefox to test it

So it is a upstream bug, use this workaround for now.

**Comment by aklinker1** [comment]
@dword-design nope, still not working. HTML pages are broken, not content scripts.

**Comment by Charan167** [comment]
@yunsii Did you find any workaround apart from that?

**Comment by yunsii** [comment]
@Charan167 Unfortunately, there is no another workaround, and I am no longer maintaining the browser extension.

**Comment by aciccarello** [comment]
The latest Firefox nightly apparently supports localhost access in MV3. I can see that the warning has changed. Unfortunately my extension is still not running properly. That might be a different compatibility issue though. 

<img width="672" height="349" alt="Screenshot of Firefox nightly extension debugging warning indicating that localhost is only permitted during development." src="https://github.com/user-attachments/assets/a5225fe8-467d-4569-8c18-1275857be418" />

Looking at the error log I wonder if web sockets are still not supported. Or maybe it's primarily the MIME type errors.

<img width="983" height="319" alt="Firefox dev tool console with errors indicating cannot connect to server at ws://localhost:3000 and Loading files was blocked due to wrong MIME type" src="https://github.com/user-attachments/assets/c76d84df-a52d-437c-8085-6f934281084d" />

**Comment by aciccarello** [comment]
I sorted out my compatibility issue and can confirm I am able to run wxt in dev mode in Firefox Nightly.

**Comment by dmose** [comment]
@aciccarello That's exciting! Is the compatibility problem something that others might run into and would be worth documenting? 

Does this mean that sufficiently new versions of Firefox and wxt (when configured to use MV3 with Firefox) will generate MV3 addons that Just Work?

**Comment by aciccarello** [comment]
@dmose The compatibility issue I had was with an old version of vite (documented in another issue).

I think before the recent Firefox change the main limitations were MV3 dev mode and android support. Now that Firefox fixed access to localhost in temporary add-ons, the first issue is fixed. There are still compatibility issues with Firefox for Android.

Note: I'm running 147.0b7 in Firefox Developer Edition and it also seems to be working. That matches what's listed in [the Firefox bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1864284). It's not yet in [the release notes](https://www.firefox.com/en-US/firefox/147.0beta/releasenotes/?redirect_source=mozilla-org&utm_source=firefox-browser&utm_medium=firefox-desktop&utm_campaign=about-dialog) but they are discussing adding it.

**Comment by aklinker1** [comment]
~~Nice, glad to hear! If someone wants to queue up a PR for this change (removing the if statement below), we can merge it once the fix is released to normal firefox.~~

https://github.com/wxt-dev/wxt/blob/3417bd3ed27d2e52e6b906393e0329565256f0e7/packages/wxt/src/core/runners/web-ext.ts#L20-L27

_edit: sorry, I think I'm still gonna mark this issue as not-planned, even though all the requirements to technically do this, Firefox MV3 is a bad UX for extension users. I'd like to keep WXT on MV2 for now. We can track adding MV3 dev mode support in #1626_

**Comment by aciccarello** [comment]
@aklinker1 Could you clarify/point to what you mean by Firefox mv3 is bad UX? What should developers be looking out for?

**Comment by aklinker1** [comment]
Granting permissions on install. Specifically, granting content scripts access to the web page. 

Sometimes Firefox just forgets to ask whether or not it can load a content script on a page and you have to go to the extension settings to manually enable it for all host permissions. 

Normal users find that very unintuitive. 

This differs from manifest V2, because it asks for all permissions on install, including host permissions for content scripts to run.

**Comment by aciccarello** [comment]
Ah, I did have someone run into that problem recently. I didn't realize that was firefox issue that it didn't ask. I was going to look into explicitly checking permissions in the extension. Thanks for explaining.

**Comment by mengxi-ream** [comment]
seems firefox support allowing connecting to local dev servers. Will we gonna have better experience? https://bugzilla.mozilla.org/show_bug.cgi?id=1864284

**Comment by aklinker1** [comment]
I'm gonna close this, I don't think we'll switch the default in the foreseeable future. We can reopen if something changes in the future.
